book/src/spec.md

```diff
@@ -0,0 +1,628 @@
+>   |     |------------|
+>   v   3 | data:  '4' |
+>         | last:  1 0 |
+>         |------------|
+>       4 | data:  '5' |
+>         | last:  0 0 |
+>         |------------|
+>       5 | data:  '6' |
+>         | last:  1 1 |
+>         '------------'
+> ```
+>
+> An argument could be made that this encodes `[[1, 2], 3, 4], [[5, 6]]`, but
+> allowing this interpretation would make the stream significantly less
+> type-safe, and there would be no way to encode non-list elements before the
+> last inner list in the outer list without introducing a `start` signal as
+> well.
+>
+> An argument could also be made that this encodes `[[1, 2]], [[3, 4, 5, 6]]`,
+> because the `last` flag delimiting `[1, 2]` falls within the first outer list
+> and the `last` flag delimiting `[3, 4, 5, 6]` falls within the second. However,
+> This would unnecessarily complicate sink logic, make manual interpretation of
+> the data structure more context-sensitive, and so on.
+>
+> One could also argue that it encodes `[[1, 2]], [[5, 6]]`, because `3` and
+> `4` are not enclosed by any inner list, but then it makes more sense to just
+> deactivate the lanes.
+>
+> Ultimately, this representation is ambiguous or at least leads to confusion,
+> and is therefore illegal.
+
+*[\\(C < 8\\)]* All last bits for lanes 0 to \\(N - 2\\) inclusive must be
+driven low by the source, and may be ignored by the sink.
+
+> The above rule ultimately means that the `last` information is transferred on
+> a transfer basis instead of on element basis, similar to AXI4-stream. This
+> can significantly decrease decoding complexity, but only allows one innermost
+> sequence to be transferred per cycle regardless of \((N\\) and the length of
+> the sequence.
+
+*[\\(C < 4\\)]* It is illegal to assert `last` bit \\(j\\) without also
+terminating all \\(j' < j\\) within the same lane.
+
+*[\\(C < 4\\)]* It is illegal to assert the `last` bit for \\(j = 0\\) when
+the respective data lane is inactive, except for empty sequences.
+
+> The above two rules mean that the `last` flags cannot be postponed.
+
+#### `stai` signal description
+
+The `stai` signal (start index) is used to deactivate data lanes with an index
+lower than the encoded binary number. It is a bit vector of size
+\\(\lceil \log_2{N} \rceil\\) to be able to encode the full index range.
+
+The state of the `stai` signal is significant only while `valid` is asserted.
+
+It is illegal to drive `stai` to the value \\(N\\) or greater.
+
+> This implies that `stai` cannot be used to disable the last lane.
+
+*[\\(C < 6\\)]* `stai` must always be driven to 0 by the source, and may be
+ignored by the sink.
+
+#### `endi` signal description
+
+The `endi` signal (end index) is used to deactivate data lanes with an index
+greater than the encoded binary number. It is a bit vector of size
+\\(\lceil \log_2{N} \rceil\\) to be able to encode the full index range.
+
+> Note that `endi` cannot be used to disable the first lane.
```

Comment:
Not sure what you mean.

Reply:
OCD 🚨
#### `stai` signal description 
and
#### `endi` signal description
have similair statements. However, these notes:

> This implies that `stai` cannot be used to disable the last lane.

> Note that `endi` cannot be used to disable the first lane.

are placed at different locations between these statements.

Reply:
That's because this constraint is placed by different parts of their spec.

Both `stai` and `endi` are defined to be treated as unsigned binary numbers. For `endi`, that means it cannot disable the first lane by definition. Per `stai`'s definition it could do that when N is not a power of two, but the later constraint saying that it is illegal to set it to N or higher makes it such that it cannot disable the last lane.

Reply:
I get it.